internal/telemetry/http_client.go

```diff
@@ -13,43 +13,72 @@ import (
 	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
 )
 
+const (
+	TransportTimeout   = 30 * time.Second
+	TransportKeepAlive = 30 * time.Second
+)
+
 var ClientProxyProductionUseErr = errors.New("HTTP client proxy can be only used in development")
 
-// HTTPClient returns newly initialized http.Client type with OpenTelemetry
-// transport. This client should be used for all REST communication on the
-// platform.
-//
-// An optional HTTP proxy can be passed in, nil means no proxy.
-func HTTPClient(ctx context.Context, proxy *url.URL) (*http.Client, error) {
-	logger := ctxval.Logger(ctx)
-	var proxyFunc func(*http.Request) (*url.URL, error)
-	if proxy != nil {
-		if config.Features.Environment != "development" {
+var (
+	sourcesTransport      *http.Transport
+	imageBuilderTransport *http.Transport
+)
+
+func init() {
+	// HTTP transports must be initialized before use
+	dialer := &net.Dialer{
+		Timeout:   TransportTimeout,
+		KeepAlive: TransportKeepAlive,
+	}
+
+	sourcesTransport = http.DefaultTransport.(*http.Transport).Clone()
+	sourcesTransport.Proxy = proxyFromConfig(func() string {
```

Comment:
This very much feels like first-class citizen already and it was very hard to parse for me what actually is happening here and I still don't get why we need to pass a function, not the string directly?
The string is not defined at this point?

Can't we just add the proxy when building the client (in `HTTP{Sources,Imagebuilder}Client`)?
It would feel much more readable and less like Proxy is first-class citizen.

(I believe we should avoid devel code, Proxy in this case, to get too rooted in the code, that's why I'd like to simlify it as much as possible even if it's slightly less performant)

Reply:
It's not nice, but we have to use functions because this is `init()` where order is unknown and configuration might not yet be available. Luckily the HTTP libraries support setting key options via functions.

The documentation/discussion in Go library indicates that HTTP transports should be cloned before any connection is made, which is guaranteed in `init()`, we do not do any HTTP requests there and should absolutely not do that.

I can move this to clients, it's just I feel like we have already quite a lot of code there and it is going to be copy-pasted around when we introduce new clients. I am aiming for the opposite - extract as much code from there to somewhere else.

Now, I understand the concern and it does not definitely look pretty. So the alternative is basically to keep the previous code, unfortunately, there is no way to get the default values (timeouts, all the numbers) from the Go library so we just need to set constants and hope for the best.